src/isomorphic/hooks/ReactComponentTreeHook.js

```diff
@@ -402,6 +402,57 @@ var ReactComponentTreeHook = {
 
   getRootIDs,
   getRegisteredIDs: getItemIDs,
+
+  pushNonStandardWarningStack(
+    isCreatingElement: boolean,
+    currentSource: ?Source,
+  ) {
+    if (typeof console.stack !== 'function') {
+      return;
+    }
+
+    var stack = [];
+    var currentOwner = ReactCurrentOwner.current;
+    var id = currentOwner && currentOwner._debugID;
+
+    try {
+      if (isCreatingElement) {
+        stack.push({
+          functionName: id ? ReactComponentTreeHook.getDisplayName(id) : null,
+          fileName: currentSource ? currentSource.fileName : null,
+          lineNumber: currentSource ? currentSource.lineNumber : null,
+        });
+      }
+
+      while (id) {
+        var element = ReactComponentTreeHook.getElement(id);
+        var parentID = ReactComponentTreeHook.getParentID(id);
+        var ownerID = ReactComponentTreeHook.getOwnerID(id);
+        var ownerName = ownerID
+          ? ReactComponentTreeHook.getDisplayName(ownerID)
+          : null;
+        var source = element && element._source;
+        stack.push({
+          functionName: ownerName,
+          fileName: source ? source.fileName : null,
+          lineNumber: source ? source.lineNumber : null,
+        });
+        id = parentID;
+      }
+    } catch (err) {
+      // Internal state is messed up.
+      // Stop building the stack (it's just a nice to have).
+    }
+
+    console.stack(stack);
```

Comment:
I was thinking that it would be good if `console.stack` indicated the root stack frame.

That way the host is welcome to add more stack frames on top of it. Essentially the same as storing `new Error().stack` and stripping off the location of the next `new Error()` before appending the custom stack.

This is useful for things like custom Promise, Algebraic Effects or call/cc implementations. E.g. Where you can to begin doing the "thenable" work you would call `console.stack` with the stack that caused the Promise to be created.

This makes sense if the `console.stack` frame is collocated with `console.warning` but that might be too limited.

So perhaps the semantics should be that any common stack frame is considered the root until the `console.endStack` frame? That seems doable.

Effectively that would mean in our case that the host environment can visualize the `warning()` and `console.warn()` call frames on top of the custom one.